libraries/AP_HAL_Linux/RCInput_SBUS.cpp

```diff
@@ -44,56 +80,103 @@ void RCInput_SBUS::set_device_path(const char *path)
-            tio.c_cflag &= ~(CSIZE | CRTSCTS | PARODD | CBAUD);
-            tio.c_cflag |= (CS8 | CSTOPB | CLOCAL | PARENB | BOTHER | CREAD);
-            tio.c_cc[VMIN] = 25;
-            tio.c_cc[VTIME] = 0;
-            ret = ioctl(fd, TCSETS2, &tio);
-            printf("SBUS ret=%d %s\n", ret, strerror(errno));
+
+    // read up to 10 frames at a time
+    uint8_t bytes[SBUS_FRAME_SIZE*10];
+    int nread;
+    
+    fd_set fds;
+    struct timeval tv;
+    
+    FD_ZERO(&fds);
+    FD_SET(fd, &fds);
+
+    tv.tv_sec = 0;
+    tv.tv_usec = 0;
+
+    // as TMIN is SBUS_FRAME_SIZE the select won't return unless there is
+    // at least SBUS_FRAME_SIZE bytes available
+    if (select(fd+1, &fds, NULL, NULL, &tv) != 1) {
+        return;
+    }
+
+#if SBUS_DEBUG_LOG
+    static int logfd = -1;
+    if (logfd == -1) {
+        logfd = open("sbus.log", O_WRONLY|O_CREAT|O_TRUNC, 0644);
+    }
+#endif
+
+#if SBUS_CAUSE_CORRUPTION
+    // deliberately lose bytes from the port
+    static unsigned corruption_counter;
+    if (corruption_counter++ % 1000 == 0) {
+        uint8_t nn = corruption_counter/1000;
+        int n2 = ::read(fd, bytes, nn);
+        dprintf(logfd, "throw %u\n", (unsigned)n2);
+    }
+#endif
+
+    // cope with there being a large number of pending bytes at
+    // the start
+    do {
+        nread = ::read(fd, bytes, sizeof(bytes));
+    } while (nread == sizeof(bytes));
+    
+    if (nread % SBUS_FRAME_SIZE != 0) {
+        /*
+          SBUS frames are 25 bytes long, and always start with
+          0x0f, but there is no other framing information to
+          prevent us getting out of sync. All we have are the
+          timing guarantees
+          
+          In this case we have read a partial frame, or lost some
+          bytes. A 25 bytes frame at 100000 baud takes 2.5ms. By
+          delaying 3.5ms here we will get any remaining bytes for
+          this frame. We shouldn't get the start of the next frame
+          as frames happen at most every 7ms
+          
+          This strategy allows us to resync even if we lose
+          bytes. It assumes an interframe gap of more than
+          3.5ms. If SBUS is run at very high rate (like 300Hz)
+          then this won't work
+        */
+        hal.scheduler->delay_microseconds(3500);
+        int n2 = ::read(fd, bytes+nread, sizeof(bytes)-nread);
+        if (n2 > 0) {
+            nread += n2;
         }
     }
-    if (fd != -1) {
-        uint8_t bytes[25];
-        int32_t nread;
-        uint8_t counter = 100;
-        do {
-            nread = ::read(fd, bytes, sizeof(bytes));
-            if (nread > 0 && bytes[0] == 0x0f) {
-                add_sbus_input(bytes, nread);
-            }
-        } while (nread == 25 && counter--);
+    if (nread % SBUS_FRAME_SIZE != 0) {
+        // if we don't have a multuple of 25 then throw away the lot
+#if SBUS_DEBUG_LOG
+        dprintf(logfd, "discard %u\n", (unsigned)nread);
+#endif
+        return;
+    }
```

Comment:
I can imagine we getting out of sync and having 25 bytes but not a frame. Could we continue checking by the starting byte?